File: integrations/mcp-experiment/proof-candidates.mjs
// Research fixture only. Neither candidate is a public Vella proof format.
import crypto from "node:crypto";
import { readFileSync } from "node:fs";
import canonicalize from "canonicalize";

const privateKey = readFileSync(new URL("../../sdk/node/test/fixtures/test-signing-private.pem", import.meta.url));
const publicKey = readFileSync(new URL("../../sdk/node/test/fixtures/test-signing-public.pem", import.meta.url));
const types = new Set(["vella-experiment/jcs", "vella-experiment/bytes"]);

// DSSE-style pre-authentication encoding binds interpretation and exact bytes.
function pae(type, bytes) {
  return Buffer.concat([Buffer.from(`DSSEv1 ${Buffer.byteLength(type)} ${type} ${bytes.length} `), bytes]);
}

export function digest(value) {
  return `sha256:${crypto.createHash("sha256").update(canonicalize(value)).digest("hex")}`;
}

export function signCandidate(value, mode = "bytes") {
  const payloadType = `vella-experiment/${mode}`;
  if (!types.has(payloadType)) throw new TypeError("unknown experimental format");
  const bytes = Buffer.from(mode === "jcs" ? canonicalize(value) : JSON.stringify(value));
  return {
    payloadType,
    payload: bytes.toString("base64"),
    signature: crypto.sign("sha256", pae(payloadType, bytes), privateKey).toString("base64"),
  };
}

export function verifyCandidate(record) {
  if (!types.has(record.payloadType)) throw new Error("unsupported experimental type");
  const bytes = Buffer.from(record.payload, "base64");
  if (!crypto.verify("sha256", pae(record.payloadType, bytes), publicKey, Buffer.from(record.signature, "base64"))) {
    throw new Error("invalid experimental signature");
  }
  const verified = JSON.parse(bytes.toString("utf8"));
  if (record.payloadType === "vella-experiment/jcs" && canonicalize(verified) !== bytes.toString("utf8")) {
    throw new Error("noncanonical experimental payload");
  }
  return verified;
}

File: integrations/mcp-experiment/proof_candidates.py
"""Disposable cross-language proof experiment, not a Vella verifier."""
import base64
import json
import sys
from pathlib import Path

import rfc8785
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec

FIXTURES = Path(__file__).resolve().parents[2] / "sdk/node/test/fixtures"
TYPES = {"vella-experiment/jcs", "vella-experiment/bytes"}


def pae(payload_type, payload):
    return f"DSSEv1 {len(payload_type.encode('utf8'))} {payload_type} {len(payload)} ".encode() + payload


request = json.load(sys.stdin)
if request["operation"] == "sign":
    if request.get("integralFloat"):
        request["value"]["integral"] = float(request["value"]["integral"])
    payload_type = "vella-experiment/" + request["mode"]
    if payload_type not in TYPES:
        raise ValueError("unknown experimental type")
    payload = rfc8785.dumps(request["value"]) if request["mode"] == "jcs" else json.dumps(
        request["value"], ensure_ascii=False, allow_nan=False, separators=(",", ":")
    ).encode("utf8")
    key = serialization.load_pem_private_key((FIXTURES / "test-signing-private.pem").read_bytes(), None)
    signature = key.sign(pae(payload_type, payload), ec.ECDSA(hashes.SHA256()))
    print(json.dumps({"payloadType": payload_type, "payload": base64.b64encode(payload).decode(), "signature": base64.b64encode(signature).decode()}))
else:
    record = request["record"]
    if record["payloadType"] not in TYPES:
        raise ValueError("unknown experimental type")
    payload = base64.b64decode(record["payload"], validate=True)
    key = serialization.load_pem_public_key((FIXTURES / "test-signing-public.pem").read_bytes())
    key.verify(base64.b64decode(record["signature"], validate=True), pae(record["payloadType"], payload), ec.ECDSA(hashes.SHA256()))
    value = json.loads(payload)
    if record["payloadType"] == "vella-experiment/jcs" and rfc8785.dumps(value) != payload:
        raise ValueError("noncanonical experimental payload")
    print(json.dumps(value))
